src/Concordium/Client/GRPC2.hs

```diff
@@ -0,0 +1,3222 @@
+queryResultCode :: QueryResult -> Int64
+queryResultCode QRInvalidArgument = -2
+queryResultCode QRInternalError = -1
+queryResultCode QRSuccess = 0
+queryResultCode QRNotFound = 1
+
+-- |If the message is given encode and write it using the provided callback.
+returnMessage ::
+    (Proto.Message (Output a), ToProto a) =>
+    (Ptr Word8 -> Int64 -> IO ()) ->
+    -- | The potential message.
+    Maybe a ->
+    IO Int64
+returnMessage copier res = case res of
+    Nothing -> return $ queryResultCode QRNotFound
+    Just v -> do
+        let encoded = Proto.encodeMessage (toProto v)
+        BS.unsafeUseAsCStringLen encoded (\(ptr, len) -> copier (castPtr ptr) (fromIntegral len))
+        return $ queryResultCode QRSuccess
+
+enqueueMessages :: (Proto.Message (Output a), ToProto a) => (Ptr Word8 -> Int64 -> IO Int32) -> [a] -> IO ThreadId
+enqueueMessages callback = enqueueProtoMessages callback . fmap toProto
+
+{- |Spawn a new thread that will invoke the provided callback on the list of
+ encoded messages. If the callback response indicates that the channel to
+ which the callback is enqueueing is full, the thread will wait. The wait time
+ follows exponential backoff strategy to a maximum of 10 seconds.
+-}
+enqueueProtoMessages :: (Proto.Message a) => (Ptr Word8 -> Int64 -> IO Int32) -> [a] -> IO ThreadId
+enqueueProtoMessages callback = forkIO . go 0 . map Proto.encodeMessage
+  where
+    go _ [] = () <$ callback nullPtr maxBound -- close the sender channel.
+    go n msgs@(msg : msgs') =
+        BS.unsafeUseAsCStringLen msg $ \(headPtr, len) -> do
+            res <- callback (castPtr headPtr) (fromIntegral len)
+            if res == 0
+                then go 0 msgs' -- reset wait time to 10ms.
+                else
+                    if res == -1
+                        then do
+                            let delay = (2 ^ n) * 10_000
+                            threadDelay delay
+                            go (min (n + 1) (10 :: Int)) msgs -- maximum delay is 10 seconds
+                        else return () -- the sender channel is now dropped, so we stop.
+
+-- V2 stuff starts here:
+
+-- |A block hash input
+data BlockHashInput = Best | LastFinal | Given BlockHash
+
+-- |An account identifier input
+type AccountIdentifierInput = AccountIdentifier
+
+instance FromProto Proto.AccountIndex where
+    type Output' Proto.AccountIndex = AccountIndex
+    fromProto = AccountIndex . deMkWord64
+instance FromProto Proto.AccountAddress where
+    type Output' Proto.AccountAddress = AccountAddress
+    fromProto = deMkSerialize
+
+instance FromProto Proto.CredentialRegistrationId where
+    type Output' Proto.CredentialRegistrationId = RawCredentialRegistrationID
+    fromProto = deMkSerialize
+instance FromProto Proto.AccountIdentifierInput where
+    type Output' Proto.AccountIdentifierInput = AccountIdentifierInput
+    fromProto a = case (a ^? ProtoFields.credId, a ^? ProtoFields.address, a ^? ProtoFields.accountIndex) of
```

Comment:
Please use pattern matching on the discriminant. Not this. This is hard to maintain.

Reply:
I agree, my question on slack was regarding this.

Could you write a quick example of what that would look like for the above case? I can not figure this out, and the documentation is a bit sparse IMO.

Reply:
```haskell
instance FromProto Proto.AccountIdentifierInput where
    type Output' Proto.AccountIdentifierInput = Maybe AccountIdentifierInput
    fromProto a = do
      aii <- a ^. ProtoFields.maybe'accountIdentifierInput
      case aii of
        Proto.AccountIdentifierInput'Address addr -> AccAddress <$> fromProto addr
        Proto.AccountIdentifierInput'CredId cid -> CredRegID <$> fromProto cid
        Proto.AccountIdentifierInput'AccountIndex ai -> Just (fromProto ai)
```

assuming you've changed `FromProto AccountAddress` to return `Maybe AccountAddress`.

Reply:
If LSP is of no help you can see the generated file with all the types in something like

```
.stack-work/dist/x86_64-linux-tinfo6/Cabal-3.6.3.0/build/autogen/Proto/V2/Concordium/Types.hs
```

It has somewhat readable documentation, for example
```haskell
{- | Fields :
     
         * 'Proto.V2.Concordium.Types_Fields.maybe'accountIdentifierInput' @:: Lens' AccountIdentifierInput (Prelude.Maybe AccountIdentifierInput'AccountIdentifierInput)@
         * 'Proto.V2.Concordium.Types_Fields.maybe'address' @:: Lens' AccountIdentifierInput (Prelude.Maybe AccountAddress)@
         * 'Proto.V2.Concordium.Types_Fields.address' @:: Lens' AccountIdentifierInput AccountAddress@
         * 'Proto.V2.Concordium.Types_Fields.maybe'credId' @:: Lens' AccountIdentifierInput (Prelude.Maybe CredentialRegistrationId)@
         * 'Proto.V2.Concordium.Types_Fields.credId' @:: Lens' AccountIdentifierInput CredentialRegistrationId@
         * 'Proto.V2.Concordium.Types_Fields.maybe'accountIndex' @:: Lens' AccountIdentifierInput (Prelude.Maybe AccountIndex)@
         * 'Proto.V2.Concordium.Types_Fields.accountIndex' @:: Lens' AccountIdentifierInput AccountIndex@ -}
data AccountIdentifierInput
  = AccountIdentifierInput'_constructor {_AccountIdentifierInput'accountIdentifierInput :: !(Prelude.Maybe AccountIdentifierInput'AccountIdentifierInput),
                                         _AccountIdentifierInput'_unknownFields :: !Data.ProtoLens.FieldSet}
  deriving stock (Prelude.Eq, Prelude.Ord)
instance Prelude.Show AccountIdentifierInput where
  showsPrec _ __x __s
    = Prelude.showChar
        '{'
        (Prelude.showString
           (Data.ProtoLens.showMessageShort __x) (Prelude.showChar '}' __s))
data AccountIdentifierInput'AccountIdentifierInput
  = AccountIdentifierInput'Address !AccountAddress |
    AccountIdentifierInput'CredId !CredentialRegistrationId |
    AccountIdentifierInput'AccountIndex !AccountIndex
  deriving stock (Prelude.Show, Prelude.Eq, Prelude.Ord)

```

Reply:
Worked like a charm!